packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -1960,7 +1977,22 @@ function dispatchAction<S, A>(
         warnIfNotCurrentlyActingUpdatesInDev(fiber);
       }
     }
-    scheduleUpdateOnFiber(fiber, lane, eventTime);
+    const root = scheduleUpdateOnFiber(fiber, lane, eventTime);
+
+    if (isTransitionLane(lane) && root !== null) {
+      let queueLanes = queue.lanes;
+
+      // If any entangled lanes are no longer pending on the root, then they
+      // must have finished. We can remove them from the
+      queueLanes = intersectLanes(queueLanes, root.pendingLanes);
+
+      // Entangle the new transition lane with the other transition lanes.
+      const newQueueLanes = mergeLanes(queueLanes, lane);
+      if (newQueueLanes !== queueLanes) {
+        queue.lanes = newQueueLanes;
+        markRootEntangled(root, newQueueLanes);
+      }
+    }
```

Comment:
I want to make sure I understand the new logic for calling `markRootEntangled`.

The only other place we do this now (I think) is in mutable source. The trigger in that case is when we detect reading from a source that has pending mutations scheduled.

I think the added calls to `markRootEntangled` are there to ensure that every _transition_ scheduled for the same state/reducer queue get entangled.

I'm not sure I understand why we need to explicitly remove already finished entangled lanes (things no longer in `pendingLanes`). I understand why we shouldn't _re-entangle them_ but I'm not sure I get why they'd still be in `queue.lanes`.

Reply:
> I understand why we shouldn't re-entangle them but I'm not sure I get why they'd still be in queue.lanes.

Yeah good question. The `lanes` field is a superset of all the transition lanes that are still pending on that queue. We don't clear it in the commit phase, we only clear it here and if the hook bails out. Neither of those heuristics will account for all cases, there will be some cases where we don't know for sure if a lane is still pending in the queue, but it's good enough for the common case, which is that there are no pending transitions on the root at all.

It's basically the same trick we use for eagerly bailing out of state updates.

Reply:
An important detail is that this is the "shared" queue, not the per-fiber queue. When we schedule updates on it, we don't know the current state. It's append-only, you can't read from it. Except with tricks like this.